File: ga/glm4_result_2024_6/result/queue_l_glm/13_queue_l.c
use std::io;
use std::error::Error;

const MIN_QUEUE_SIZE: usize = 5;

type ElementType = i32;

struct Queue {
    capacity: usize,
    front: usize,
    rear: usize,
    size: usize,
    array: Vec ElementType,
}

impl Queue {
    fn new(capacity: usize) -> Queue {
        if capacity < MIN_QUEUE_SIZE {
            panic!("Queue size is too small");
        }
        Queue {
            capacity: capacity,
            front: 0,
            rear: 0,
            size: 0,
            array: vec![0; capacity],
        }
    }

    fn is_empty(&self) -> bool {
        self.size == 0
    }

    fn is_full(&self) -> bool {
        self.size == self.capacity
    }

    fn enqueue(&mut self, value: ElementType) {
        if self.is_full() {
            panic!("Full queue");
        }
        self.size += 1;
        self.array[self.rear] = value;
        self.rear = (self.rear + 1) % self.capacity;
    }

    fn front(&self) -> ElementType {
        if self.is_empty() {
            panic!("Empty queue");
        }
        self.array[self.front]
    }

    fn dequeue(&mut self) {
        if self.is_empty() {
            panic!("Empty queue");
        }
        self.size -= 1;
        self.front = (self.front + 1) % self.capacity;
    }

    fn front_and_dequeue(&mut self) -> ElementType {
        let value = self.front();
        self.dequeue();
        value
    }
}

fn main() {
    let mut queue: Queue;
    let mut queue_length: usize;

    loop {
        println!("Enter the queue length (must be at least {}):", MIN_QUEUE_SIZE);
        match read_input() {
            Ok(length) => {
                queue_length = length;
                if queue_length < MIN_QUEUE_SIZE {
                    println!("Queue size is too small, please try again.");
                } else {
                    break;
                }
            }
            Err(e) => {
                println!("Invalid input: {}", e);
            }
        }
    }

    queue = Queue::new(queue_length);

    // Test: Attempt to dequeue from an empty queue
    if !queue.is_empty() {
        queue.dequeue();
    } else {
        println!("Queue is empty, cannot dequeue.");
    }

    // Enqueue elements and check if the queue is full
    for i in 0..queue_length - 1 {
        if queue.is_full() {
            println!("Queue is full");
        } else {
            queue.enqueue(i);
        }
    }

    // Test: Attempt to enqueue to a full queue
    if !queue.is_full() {
        println!("Enter a value to enqueue:");
        match read_input() {
            Ok(value) => {
                queue.enqueue(value);
            }
            Err(e) => {
                println!("Invalid input: {}", e);
            }
        }
    } else {
        println!("Queue is full, cannot enqueue.");
    }

    // Use FrontAndDequeue to get and remove elements
    while !queue.is_empty() {
        println!("FrontAndDequeue: {}", queue.front_and_dequeue());
    }

    // Enqueue elements again
    for i in 0..queue_length - 2 {
        queue.enqueue(i);
    }

    // Use Front and Dequeue
    while !queue.is_empty() {
        println!("Front: {}", queue.front());
        queue.dequeue();
    }
}

fn read_input() -> Result<usize, Box<dyn Error>> {
    let mut input = String::new();
    io::stdin().read_line(&mut input)?;
    input.trim().parse::<usize>()
}

File: ga/4omini_result_2024_7/result/queue_l_mini/13_queue_l.rs
use std::error::Error;
use std::fmt;

const MIN_QUEUE_SIZE: usize = 5;

#[derive(Debug)]
struct Queue {
    capacity: usize,
    front: usize,
    rear: usize,
    size: usize,
    array: Vec<i32>, // Assuming the ElementType is Integer
}

#[derive(Debug)]
struct QueueError(String);

impl fmt::Display for QueueError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "QueueError: {}", self.0)
    }
}

impl Error for QueueError {}

impl Queue {
    fn new(max_elements: usize) -> Result<Queue, Box<dyn Error>> {
        if max_elements < MIN_QUEUE_SIZE {
            return Err(Box::new(QueueError("Queue size is too small".to_string())));
        }

        Ok(Queue {
            capacity: max_elements,
            front: 0,
            rear: max_elements - 1,
            size: 0,
            array: vec![0; max_elements],
        })
    }

    fn is_empty(&self) -> bool {
        self.size == 0
    }

    fn is_full(&self) -> bool {
        self.size == self.capacity
    }

    fn enqueue(&mut self, x: i32) -> Result<(), Box<dyn Error>> {
        if self.is_full() {
            return Err(Box::new(QueueError("Full queue".to_string())));
        }

        self.size += 1;
        self.rear = (self.rear + 1) % self.capacity;
        self.array[self.rear] = x;
        Ok(())
    }

    fn front(&self) -> Result<i32, Box<dyn Error>> {
        if self.is_empty() {
            return Err(Box::new(QueueError("Empty queue".to_string())));
        }
        Ok(self.array[self.front])
    }

    fn dequeue(&mut self) -> Result<(), Box<dyn Error>> {
        if self.is_empty() {
            return Err(Box::new(QueueError("Empty queue".to_string())));
        }

        self.size -= 1;
        self.front = (self.front + 1) % self.capacity;
        Ok(())
    }

    fn front_and_dequeue(&mut self) -> Result<i32, Box<dyn Error>> {
        if self.is_empty() {
            return Err(Box::new(QueueError("Empty queue".to_string())));
        }

        let value = self.array[self.front];
        self.size -= 1;
        self.front = (self.front + 1) % self.capacity;
        Ok(value)
    }

    fn make_empty(&mut self) {
        self.size = 0;
        self.front = 0;
        self.rear = self.capacity - 1;
    }
}

fn main() -> Result<(), Box<dyn Error>> {
    use std::io;
    use std::num::ParseIntError;

    let mut input = String::new();

    print!("Enter queue length: ");
    io::stdin().read_line(&mut input)?;
    let queue_length: usize = input.trim().parse().map_err(|e: ParseIntError| Box::new(e))?;

    let mut queue = Queue::new(queue_length)?;

    // Test: Attempt to dequeue from an empty queue
    if !queue.is_empty() {
        queue.dequeue()?;
    } else {
        println!("Queue is empty, cannot dequeue.");
    }

    // Enqueue elements and check if the queue is full
    for i in 0..(queue_length - 1) {
        if queue.is_full() {
            println!("Queue is full");
        } else {
            queue.enqueue(i)?;
        }
    }

    // Test: Attempt to enqueue to a full queue
    if !queue.is_full() {
        input.clear();
        print!("Enter an enqueue value: ");
        io::stdin().read_line(&mut input)?;
        let enqueue_value: i32 = input.trim().parse().map_err(|e: ParseIntError| Box::new(e))?;
        queue.enqueue(enqueue_value)?;
    } else {
        println!("Queue is full, cannot enqueue.");
    }

    // Use FrontAndDequeue to get and remove elements
    while !queue.is_empty() {
        println!("FrontAndDequeue: {}", queue.front_and_dequeue()?);
    }

    // Enqueue elements again
    for i in 0..(queue_length - 2) {
        queue.enqueue(i)?;
    }

    // Use Front and Dequeue
    while !queue.is_empty() {
        println!("Front: {}", queue.front()?);
        queue.dequeue()?;
    }

    Ok(())
}